introspection-engine/core/src/rpc.rs

```diff
@@ -66,85 +69,77 @@ impl RpcImpl {
-            }
-            Ok(introspection_result) => {
-                let warnings = match could_not_parse_input_data_model {
-                    true if reintrospect => {
-                        let mut warnings = introspection_result.warnings;
-                        warnings.push(Warning {
-                        code: 0,
-                        message:
-                        "The input datamodel could not be parsed. This means it was not used to enrich the introspected datamodel with previous manual changes."
-                            .into(),
-                        affected: serde_json::Value::Null,
-                    });
-                        warnings
-                    }
-                    _ => introspection_result.warnings,
-                };
-
-                let result = IntrospectionResultOutput {
-                    datamodel: datamodel::render_datamodel_and_config_to_string(
-                        &introspection_result.datamodel,
-                        &config,
-                    )
-                    .map_err(Error::from)?,
-                    warnings,
-                    version: introspection_result.version,
-                };
-
-                Ok(result)
-            }
-            Err(e) => Err(render_jsonrpc_error(Error::from(e))),
-        }
+        result.map_err(|e| render_jsonrpc_error(e, &schema))
     }
 
     pub async fn list_databases_internal(schema: String) -> RpcResult<Vec<String>> {
-        let connector = RpcImpl::load_connector(&schema).await?;
-        Ok(connector.list_databases().await.map_err(Error::from)?)
+        let (_, _, connector) = RpcImpl::load_connector(&schema)
+            .await
+            .map_err(|e| render_jsonrpc_error(e, &schema))?;
+        Ok(connector
+            .list_databases()
+            .await
+            .map_err(|e| render_jsonrpc_error(Error::from(e), &schema))?)
     }
 
     pub async fn get_database_description(schema: String) -> RpcResult<String> {
-        let connector = RpcImpl::load_connector(&schema).await?;
-        Ok(connector.get_database_description().await.map_err(Error::from)?)
+        let (_, _, connector) = RpcImpl::load_connector(&schema)
+            .await
+            .map_err(|e| render_jsonrpc_error(e, &schema))?;
+        Ok(connector
+            .get_database_description()
+            .await
+            .map_err(|e| render_jsonrpc_error(Error::from(e), &schema))?)
     }
 
     pub async fn get_database_metadata_internal(schema: String) -> RpcResult<DatabaseMetadata> {
-        let connector = RpcImpl::load_connector(&schema).await?;
-        Ok(connector.get_metadata().await.map_err(Error::from)?)
+        let (_, _, connector) = RpcImpl::load_connector(&schema)
+            .await
+            .map_err(|err| render_jsonrpc_error(err, &schema))?;
+        Ok(connector
+            .get_metadata()
+            .await
+            .map_err(|e| render_jsonrpc_error(Error::from(e), &schema))?)
```

Comment:
In this function and the two above, I think we could alleviate the `map_err` boilerplate by either:

- Wrapping the body in an async block, then await and `map_err` on that.
- (preferred solution) do something like `catch` in the sql migration connector: a function that takes the schema and an async block, runs the async block and does the error handling.

Reply:
Went with the second one.